cub/agent/agent_reduce.cuh

```diff
@@ -27,357 +27,418 @@
     enum
     {
-        BLOCK_THREADS       = AgentReducePolicy::BLOCK_THREADS,
-        ITEMS_PER_THREAD    = AgentReducePolicy::ITEMS_PER_THREAD,
-        VECTOR_LOAD_LENGTH  = CUB_MIN(int(ITEMS_PER_THREAD), int(AgentReducePolicy::VECTOR_LOAD_LENGTH)),
-        TILE_ITEMS          = BLOCK_THREADS * ITEMS_PER_THREAD,
-
-        // Can vectorize according to the policy if the input iterator is a native pointer to a primitive type
-        ATTEMPT_VECTORIZATION   = (VECTOR_LOAD_LENGTH > 1) &&
-                                    (ITEMS_PER_THREAD % VECTOR_LOAD_LENGTH == 0) &&
-                                    (std::is_pointer<InputIteratorT>::value) && Traits<InputT>::PRIMITIVE,
-
+      WORDS = ITEMS_PER_THREAD / VECTOR_LOAD_LENGTH
     };
 
-    static const CacheLoadModifier    LOAD_MODIFIER   = AgentReducePolicy::LOAD_MODIFIER;
-    static const BlockReduceAlgorithm BLOCK_ALGORITHM = AgentReducePolicy::BLOCK_ALGORITHM;
-
-    /// Parameterized BlockReduce primitive
-    using BlockReduceT =
-      BlockReduce<OutputT, BLOCK_THREADS, AgentReducePolicy::BLOCK_ALGORITHM>;
-
-    /// Shared memory type required by this thread block
-    struct _TempStorage
+    // Fabricate a vectorized input iterator
+    InputT *d_in_unqualified = const_cast<InputT *>(d_in) + block_offset +
+                               (threadIdx.x * VECTOR_LOAD_LENGTH);
+    CacheModifiedInputIterator<AgentReducePolicy::LOAD_MODIFIER, VectorT, OffsetT>
+      d_vec_in(reinterpret_cast<VectorT *>(d_in_unqualified));
+
+    // Load items as vector items
+    InputT input_items[ITEMS_PER_THREAD];
+    VectorT *vec_items = reinterpret_cast<VectorT *>(input_items);
+#pragma unroll
+    for (int i = 0; i < WORDS; ++i)
     {
-        typename BlockReduceT::TempStorage  reduce;
-    };
-
-    /// Alias wrapper allowing storage to be unioned
-    struct TempStorage : Uninitialized<_TempStorage> {};
-
-
-    //---------------------------------------------------------------------
-    // Per-thread fields
-    //---------------------------------------------------------------------
-
-    _TempStorage&           temp_storage;       ///< Reference to temp_storage
-    InputIteratorT          d_in;               ///< Input data to reduce
-    WrappedInputIteratorT   d_wrapped_in;       ///< Wrapped input data to reduce
-    ReductionOp             reduction_op;       ///< Binary reduction operator
-
-
-    //---------------------------------------------------------------------
-    // Utility
-    //---------------------------------------------------------------------
-
-
-    // Whether or not the input is aligned with the vector type (specialized for types we can vectorize)
-    template <typename Iterator>
-    static __device__ __forceinline__ bool IsAligned(
-        Iterator        d_in,
-        Int2Type<true>  /*can_vectorize*/)
-    {
-        return (size_t(d_in) & (sizeof(VectorT) - 1)) == 0;
+      vec_items[i] = d_vec_in[BLOCK_THREADS * i];
     }
```

Comment:
No change requested: That seems like a really common operation. Should we have a helper function for it?

Reply:
I think we should, once this is used in more than one place.